odinochka.js

```diff
@@ -15,22 +15,63 @@ function newTabs(data) {
     data.tabs.forEach(o => chrome.tabs.create({url: o.url, pinned: o.pinned}))
 }
 
-function cssfilter(x) {
-    let node = document.getElementById("cssfilterstyle");
-    if(x.target.value != "") {
-        node.innerHTML = `a.tab:not([href*="${x.target.value}"]) {display:none} `;
-        // TODO someday when :has works, also hide the empty groups
-    }
-    else {
-        node.innerHTML = "";
+function debounce(func, wait, immediate) {
+    let timeout;
+    return function(e) {
+        const context = e,
+            args = arguments;
+        const later = () => {
+            timeout = null;
+            if (!immediate) func.apply(context, args);
+        };
+        const callNow = immediate && !timeout;
+        clearTimeout(timeout);
+        timeout = setTimeout(later, wait);
+        if (callNow) func.apply(context, args);
+    };
+}
+
+function filterResults(e) {
+    // render entire list from indexDB if input is empty
+    const query = e.target.value
+    if (query === '') {
+        render()
+        return
     }
+    window.indexedDB.open("odinochka", 5).onsuccess = function(event){
+        let db = event.target.result;
+
+        let tx = db.transaction('tabgroups', 'readonly');
+        let store = tx.objectStore('tabgroups');
+
+        let allRecords = store.getAll()
+        allRecords.onsuccess = function() {
+            const filteredResults = allRecords.result
+                .map(function(cursor) {
+                    const filteredTabs = cursor.tabs
+                        .map(function(tab) {
+                            if (tab.title.toLowerCase().match(query)
+                                || tab.url.toLowerCase().match(query)) {
+                                return tab
+                            }
+                        })
+                        .filter(function(tab) { return tab !== undefined})
+                    if (filteredTabs.length) {
+                        cursor.tabs = filteredTabs
+                        return cursor
+                    }
+                })
+                .filter(function(cursor) { return cursor !== undefined })
```

Comment:
@nfultz i believe to improve the time complexity of this filtering logic - you will need to play with the format of the data so you can do key lookup as much as possible instead of looping over arrays.

Reply:
I agree, but it isn't quite straightforward, because IndexDB only lets you scan prefixes of the key, so matching text in the middle is a challenge. One workaround is to scan the indexes of url and title to find matching groups, then fetch the groups and filter down the matches. Going to meditate on it.

Reply:
right - this would not filter group names at the moment as written